lib/movable_ink/aws/ec2.rb

```diff
@@ -75,34 +76,84 @@ def me
         @me ||= all_instances.select{|instance| instance.instance_id == instance_id}.first rescue nil
       end
 
-      def instances(role:, exclude_roles: [], region: my_region, availability_zone: nil, exact_match: false, use_cache: true)
-        roles = role.split(/\s*,\s*/)
-        if use_cache == false
-          filter = default_filter.push({
-            name: 'tag:mi:roles',
-            values: roles
-          })
-          instances = load_all_instances(region, filter: filter)
-        else
-          instances = all_instances(region: region).select { |instance|
-            instance.tags.select{ |tag| tag.key == 'mi:roles' }.detect { |tag|
-              tag_roles = tag.value.split(/\s*,\s*/)
-              if exact_match
-                tag_roles == roles
-              else
-                exclude_roles.push('decommissioned')
-                tag_roles.any? { |tag_role| roles.include?(tag_role) } && !tag_roles.any? { |role| exclude_roles.include?(role) }
-              end
+      def instances(role:, exclude_roles: [], region: my_region, availability_zone: nil, exact_match: false, use_cache: true, discovery_type: 'ec2')
+        if discovery_type == 'ec2'
```

Comment:
Can you create two new methods for each discovery type and extract the implementation?

Reply:
Refactored this as requested.